### PR TITLE
Image regeneration and thumbnail improvements

### PR DESCRIPTION
**Pull request by mikejolley**
This PR is a combination of both:

- https://github.com/woocommerce/woocommerce/pull/18850 which improves checks before on the fly generation and background processing to work with CDNs
- https://github.com/woocommerce/woocommerce/pull/18947 Gallery thumbnail size used for the square crop thumbs below product page image galleries.

Notes:

- Background processing is on by default for regular stores but not MU.
- When background processing is running an admin notice will be shown and allow you to cancel the job.
- Turning it off with the filter will cancel current jobs.
- There is a tool to trigger regen.
- Regen will fire less often; on customizer publish and theme swtich but only if the sizes change.
- If using Jetpack Photon, bg regeneration is disabled. Let Photon handle it.

### CODE REVIEW

File: includes/api/class-wc-rest-system-status-tools-controller.php [hunk 1]
```diff
@@ -179,6 +179,11 @@ public function get_tools() {
 				'button' => __( 'Reset', 'woocommerce' ),
 				'desc'   => __( 'This will reset your usage tracking settings, causing it to show the opt-in banner again and not sending any data.', 'woocommerce' ),
 			),
+			'regenerate_thumbnails'      => array(
+				'name'   => __( 'Regenerate shop thumbnails', 'woocommerce' ),
+				'button' => __( 'Regenerate', 'woocommerce' ),
+				'desc'   => __( 'This will regenerate all shop thumbnails to match your theme and/or image setting.', 'woocommerce' ),
```
Comment (kloon) [inline includes/api/class-wc-rest-system-status-tools-controller.php]: should be settings not setting

**Review by kloon — CHANGES_REQUESTED** [review]
Just one small typo I spotted while going through the code. Still busy testing functionality.

**Comment by kloon** [comment]
Works for me, tested with and without CDN and both times it was working correctly. Might be work getting a few more testers on this just for certainty.

**Comment by horaciotorrent** [comment]
Hi, i have de Master Study

We used to have three image sizes:
– Single: for the product page
– Catalog: for the shop and archives
– Thumbnail: for the image gallery below the main product image

My shop (with Master Study).
How can I define the width and also the height of the Catalog image (interior product) and what is cut to that extent?

Regards

File: includes/class-wc-regenerate-images.php [hunk 2]
```diff
@@ -23,27 +23,88 @@ class WC_Regenerate_Images {
...
+		// Regenerate thumbnails in the background after settings change. Not ran on multisite to avoid multiple simultanious jobs..
+		if ( apply_filters( 'woocommerce_background_image_regeneration', ! is_multisite() ) ) {
+			include_once WC_ABSPATH . 'includes/class-wc-regenerate-images-request.php';
+
+			self::$background_process = new WC_Regenerate_Images_Request();
+
+			add_action( 'admin_init', array( __CLASS__, 'regenerating_notice' ) );
+			add_action( 'woocommerce_hide_regenerating_thumbnails_notice', array( __CLASS__, 'dismiss_regenerating_notice' ) );
+			add_action( 'customize_save_after', array( __CLASS__, 'maybe_regenerate_images' ) );
+			add_action( 'after_switch_theme', array( __CLASS__, 'maybe_regenerate_images' ) );
```
Comment (claudiulodro) [inline includes/class-wc-regenerate-images.php]: I get notices when switching themes: `Deprecated: Non-static method WC_Regenerate_Images::maybe_regenerate_images() should not be called statically in /srv/www/woodev/wp-includes/class-wp-hook.php on line 288`.

**Comment by claudiulodro** [comment]
Works well (just a notice needs fixing). Regenerate button was a good idea. 👍

**Comment by JulienMelissas** [comment]
Thanks for all the work on this folks! I'm going to try and find some time tonight to test with my CDN.

File: includes/class-wc-regenerate-images.php (hunk 2 above)
Comment (horaciotorrent) [reply includes/class-wc-regenerate-images.php]: I do not have the class-wc-regenerate-images.php file in my wp-include folder :-(

**Comment by WPprodigy** [comment]
Some initial thoughts:

> an admin notice will be shown and allow you to cancel the job.

The cancel button should stick out a bit more. Looks rather close to a dismiss button at the moment. Would maybe even consider a prompt that makes sure the user knows the dangers of doing so.

> There is a tool to trigger regen.

Double notices are confusing: http://cld.wthms.co/tBcJd7. Need to prevent users from just refreshing this page, starting the regen process over and over again xD

> woocommerce_thumbnail_2x

`woocommerce_thumbnail_2x` was removed from the list of images to regenerate. Talked about this a little, but just wanted to note here s it's noticed.

**Comment by mikejolley** [comment]
Not worried about the dismiss/cancel button now it's reworded. I think it's pretty explicit.

The double notice has been reworded so it's not as confusing if it does occur.

Last thing to look into is Jetpack CDN.

**Comment by claudiulodro** [comment]
The code for this looks good as far as I can tell and the new changes look reasonable. However, I no longer get image regeneration for custom aspect ratios on my local environment (or custom aspect ratios are not getting applied?). No matter what aspect ratio I choose for my thumbnails it looks good in the customizer but when published the image ratios in my shop stay the same. I've tried regenerating the thumbnails using the button and switching themes but no dice. It definitely worked well in earlier versions of this PR.

**Pull request merged**